app/WebParsing/UtsgJsonParser.hs

```diff
@@ -42,17 +31,53 @@ insertAllMeetings org = do
     manager <- liftIO $ newManager tlsManagerSettings
     response <- liftIO $ httpLbs request' manager
     let respBody = responseBody response
+    return respBody
 
-    -- decode the response
+-- Get the page number, page size and total number of courses from response
+getPageInfo :: ByteString -> Maybe (Int, Int, Int)
+getPageInfo respBody = do
+    json <- decode respBody
+    flip parseMaybe json $ \obj -> do
+      payload <- obj .: "payload"
+      pageableCourse <- payload .: "pageableCourse"
+      page <- pageableCourse .: "page"
+      pageSize <- pageableCourse .: "pageSize"
+      totalCourses <- pageableCourse .: "total"
+      return (page, pageSize, totalCourses)
+          
+-- Helper function to insert courses for a page of a HTTP response
+insertCourses :: ByteString -> SqlPersistM ()
+insertCourses respBody = do
     let meetings :: Maybe DBList = decode respBody
     case meetings of
       Nothing -> return ()
-      Just dblist ->  mapM_ insertMeeting $ flattenDBList dblist
+      Just dblist -> mapM_ insertMeeting $ flattenDBList dblist
 
 -- | Helper function to flatten the list of DB Objects
 flattenDBList :: DBList -> [MeetTime]
 flattenDBList (DBList meetings) = concatMap (\(DB meetTimes) -> meetTimes) meetings
 
+-- | insert/update all the data into the Meeting and Times schema by creating and sending
+--   the http request to Artsci Timetable and then parsing the JSON response
+insertAllMeetings :: Int -> SqlPersistM ()
+insertAllMeetings page = do
+    respBody <- liftIO $ makeRequest page
+
+    let pageInfo :: Maybe (Int, Int, Int) = getPageInfo respBody
+    case pageInfo of 
+      Nothing -> return ()
+      Just (page', pageSize, totalCourses) -> do
+        if (page' - 1) * pageSize > totalCourses  -- base case
+        then liftIO $ putStrLn "All courses have been parsed."
+        else do
+          liftIO $ print $ "Parsing results for page " ++ show page' ++ " of " ++ show totalPages
```

Comment:
Overall I think the logic isn't quite right. Given an input `page`, the request (`makeRequest page`) should always be made, and then the resulting message should be printed and courses should always be inserted. The only thing an if statement should be used determine is whether to make a *new* request or print the "All courses have been parsed" message.